bundle/edu.gemini.spModel.io/src/main/scala/edu/gemini/spModel/io/impl/migration/to2016B/To2016B.scala

```diff
@@ -68,4 +81,116 @@ object To2016B extends Migration {
       Option(paramSet.getParamSet("guideEnv")).fold(addEmptyGuideEnv(paramSet))(addDisabledAutomaticGroup)
     }
   }
+
+  // Add a scheduling block to nonsidereal observations that don't have one. A duration of 0ms means
+  // that the planned time will be used by internal calculations, which is a reasonable default.
+  def updateSchedulingBlocks(d: Document): Unit =
+    for {
+      (o, b) <- obsAndBases(d) if o.value("schedulingBlockStart").isEmpty
+      date   <- b.value("validAt").map(SPTargetPio.parseDate)
+    } {
+      Pio.addLongParam(fact, o, "schedulingBlockStart", date.getTime)
+      Pio.addLongParam(fact, o, "schedulingBlockDuration", 0L)
+    }
+
+  // Update targets to the new model
+  def updateTargets(d: Document): Unit =
+    allTargets(d).foreach { t =>
+
+      // In order to determine the target type we need the system as well as the coordinates
+      // because a TOO target is encoded as a sidereal target at the origin.
+      val data: Option[(String, Coordinates)] =
+        for {
+          system <- t.value("system")
+          ra     <- t.value("c1").flatMap(s => (Angle.parseHMS(s) orElse Angle.parseDegrees(s)).toOption).map(RightAscension.fromAngle)
+          dec    <- t.value("c2").flatMap(s => (Angle.parseDMS(s) orElse Angle.parseDegrees(s)).toOption).flatMap(Declination.fromAngle)
+        } yield (system, Coordinates(ra, dec))
+
+      // Construct a new target
+      val newTarget: Target =
+        data.map {
+          case ("J2000", Coordinates.zero) => TooTarget.empty
+          case ("J2000",               cs) => sidereal(t, cs).exec(SiderealTarget.empty)
+          case ("JPL minor body",      cs) => nonsidereal(t, cs).exec(NonSiderealTarget.empty)
+          case ("MPC minor planet",    cs) => nonsidereal(t, cs).exec(NonSiderealTarget.empty)
+          case ("Solar system object", cs) => (nonsidereal(t, cs) *> named(t)).exec(NonSiderealTarget.empty)
+        }.map(common(t).exec) getOrElse sys.error("Can't recognize target.")
+
+      // Drop it into the paramset.
+      // TODO: remove everything else!
+      t.addParamSet(TargetParamSetCodecs.TargetParamSetCodec.encode("target", newTarget))
+
+    }
+
+  // Properties common to all target types
+  def common(ps: ParamSet): State[Target, Unit] =
+    for {
+      _ <- Target.name                ?:= ps.value("name")
```

Comment:
Yeah `:=` turns a lens into a state action:  `Lens[A,B] := B` gives you a `State[A, B]`. So this lets you use imperative-looking code to write a pure functional series of lens transformations. So I compose all the "assignments" into a single `State` program and then use `exec` to apply it to an `A`.

`?:=` is defined up at the top of the file. It does the same thing but for `Lens[A,B] := Option[B]` that is a no-op if the value is empty.